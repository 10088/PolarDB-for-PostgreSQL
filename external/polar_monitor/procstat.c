/*-------------------------------------------------------------------------
 *
 * procstat.c
 *
 *	Copyright (c) 2020, Alibaba.inc
 *
 *	IDENTIFICATION
 *		external/polar_monitor/procstat.c
 *
 *-------------------------------------------------------------------------
 */

#include <stdlib.h>
#include <string.h>
#include "procstat.h"

#define MAXPATHLENGTH 80
#define MAXSTATLEN    1024

static const char * delim = " ";
static char path[MAXPATHLENGTH];
static char procbuf[MAXSTATLEN];

static char * readone(char* curstr, void *x, PROCARGTYPE argtype) ;
static int parse_proc_pid_stat(int pid, polar_proc_stat *prostat);
static int parse_proc_pid_statm(int pid, polar_proc_stat *stat);

static char * readone(char* curstr, void *x, PROCARGTYPE argtype) 
{
	char* token = NULL;
	char* savestr = NULL;

	token = strtok_r (curstr, delim, &savestr);
	if(token)
	{
		switch (argtype)
		{
		case PROCINT : 
			*(int *)x = (int)atoi(token);
			break;

		case PROCNUM : 
			*(num *)x = (num)atoll(token);
			break;

		case PROCL : 
			*(long *)x = (long)atol(token);
			break;

		case PROCLL :
			*(long long *)x = (long long)atoll(token);
			break;

		case PROCSTR :
			strncpy((char *)x, token, savestr-curstr);
			break;
		
		default:
			break;
		}
	}
	return savestr;
}

/*  
 *  POLAR:
 * 	Parse a proc/[pid]/stat file 
 *  if return 1 ,not find stat file
 *  return 0, successful!
 */
static int
parse_proc_pid_stat(int pid, polar_proc_stat *prostat)
{
	char * curstr = NULL;
	char tty_null[64];
	FILE *input;
	int 	i;

    input = NULL;
    memset(path, 0 ,sizeof(path));
    sprintf(path,"/proc/%d/stat",pid);

    input = fopen(path, "r");
    if(!input) {
      return 1;
    }

    if(!fread(procbuf, MAXSTATLEN, 1, input))
	{
		if(ferror(input))
		{
			fclose(input);
			return 1;
  		}
	}

    curstr = readone(procbuf, tty_null, PROCSTR);
    curstr = readone(curstr, tty_null, PROCSTR);

	/* We don't need extra here */
	for (i = 0; i < 11; i++)
		curstr = readone(curstr, tty_null, PROCSTR);

    curstr = readone(curstr, &prostat->utime, PROCNUM);
    curstr = readone(curstr, &prostat->stime, PROCNUM);

    return fclose(input);
}

/*  
 *  POLAR:
 * 	Parse a proc/[pid]/statm file 
 *  if return 1 ,not find stat file
 *  return 0, successful!
 */
static int
parse_proc_pid_statm(int pid, polar_proc_stat *stat)
{
	char * curstr = NULL;
	char tty_null[64];
	FILE *input;

    input = NULL;
    memset(path, 0 ,sizeof(path));
    sprintf(path,"/proc/%d/statm",pid);

    input = fopen(path, "r");
    if(!input) {
      return 1;
    }

    if(!fread(procbuf, MAXSTATLEN, 1, input))
	{
		if(ferror(input))
		{
			fclose(input);
			return 1;
  		}
	}

	curstr = readone(procbuf, tty_null, PROCSTR); 
	curstr = readone(curstr, &stat->rss, PROCNUM);
	curstr = readone(curstr, &stat->share, PROCNUM);
	
    return fclose(input);
}

int 
polar_get_proc_stat(int pid, polar_proc_stat *stat)
{
	int ret = parse_proc_pid_stat(pid, stat);
	if (ret)
		return ret;

	return parse_proc_pid_statm(pid, stat);
}