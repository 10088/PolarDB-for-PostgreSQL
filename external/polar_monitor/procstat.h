/*-------------------------------------------------------------------------
 *
 * procstat.h
 *
 *	Copyright (c) 2020, Alibaba.inc
 *
 * IDENTIFICATION
 *		external/polar_monitor/procstat.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef PROCSTAT_H
#define PROCSTAT_H

#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include <linux/limits.h>
#include <sys/times.h>

typedef enum PROCARGTYPE{
	PROCNUM = 0, 
	PROCSTR   =  1 ,
    PROCL = 2,
    PROCLL = 3,
    PROCULL = 4,
    PROCINT = 5
}PROCARGTYPE;

typedef unsigned long int num;

typedef struct polar_proc_stat
{
    num utime;                              /** user mode jiffies **/
    num stime;                             /** kernel mode jiffies **/
    num rss;                                /** RssAnon + RssFile + RssShare **/
    num share;                              /** RssFile + RssShare */
} polar_proc_stat;

extern int polar_get_proc_stat(int pid, polar_proc_stat *stat);
#endif							/* PROCSTAT_H */
