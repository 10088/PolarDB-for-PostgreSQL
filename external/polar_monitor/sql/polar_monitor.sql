-- Test access privileges
--
--setup
create extension polar_monitor;

-- buffer monitor
select COUNT(*) >= 0 AS result from polar_normal_buffercache;
select COUNT(*) >= 0 AS result from polar_copy_buffercache;
select COUNT(polar_flushlist()) >= 0 As result;
select COUNT(polar_cbuf()) >= 0 As result;

--cleanup
drop extension polar_monitor;
