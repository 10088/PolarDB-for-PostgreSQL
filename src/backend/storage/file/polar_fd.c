/*-------------------------------------------------------------------------
 *
 * polar_fd.c
 *
 *	  PolarDB Virtual file descriptor code.
 *
 *	  Copyright (c) 2020, Alibaba Group Holding Limited
 *	  Licensed under the Apache License, Version 2.0 (the "License");
 *	  you may not use this file except in compliance with the License.
 *	  You may obtain a copy of the License at
 *	  http://www.apache.org/licenses/LICENSE-2.0
 *	  Unless required by applicable law or agreed to in writing, software
 *	  distributed under the License is distributed on an "AS IS" BASIS,
 *	  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *	  See the License for the specific language governing permissions and
 *	  limitations under the License.
 *
 * IDENTIFICATION
 *	  src/backend/storage/file/polar_fd.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include <sys/file.h>
#include <sys/param.h>
#include <sys/stat.h>
#ifndef WIN32
#include <sys/mman.h>
#endif
#include <limits.h>
#include <unistd.h>
#include <fcntl.h>
#ifdef HAVE_SYS_RESOURCE_H
#include <sys/resource.h>		/* for getrlimit */
#endif

#include "miscadmin.h"
#include "access/xact.h"
#include "access/xlog_internal.h"
#include "access/xlog.h"
#include "catalog/pg_tablespace.h"
#include "common/file_perm.h"
#include "pgstat.h"
#include "portability/mem.h"
#include "storage/fd.h"
#include "storage/ipc.h"
#include "utils/guc.h"
#include "utils/resowner_private.h"
#include "storage/polar_fd.h"

bool	polar_mount_pfs_readonly_mode = true;
bool	polar_openfile_with_readonly_in_replica = false;
int		polar_vfs_switch = POLAR_VFS_SWITCH_LOCAL;

static inline const vfs_mgr* polar_get_local_vfs_mgr(const char *path);

/*
 * POLAR: The virtual file interface
 * 1 The standard file interface function is used by default.
 * polar_vfs_switch = POLAR_VFS_SWITCH_LOCAL
 * 2 When a storage plugin is accessed, all io call implemented in the plugin.
 * polar_vfs_switch = POLAR_VFS_SWITCH_PLUGIN
 * 3 polar_vfs means use function in plugin to read write data which include local file or 
 * stared storage file.
 */
vfs_mgr polar_vfs[] =
{
	{
		.vfs_mount = NULL,
		.vfs_umount = NULL,
		.vfs_remount = NULL,
		.vfs_open = (vfs_open_type)open,
		.vfs_creat = creat,
		.vfs_close = close,
		.vfs_read = read,
		.vfs_write = write,
		.vfs_pread = pread,
		.vfs_pwrite = pwrite,
		.vfs_stat = stat,
		.vfs_fstat = fstat,
		.vfs_lstat = lstat,
		.vfs_lseek = lseek,
		.vfs_access = access,
		.vfs_fsync = pg_fsync,
		.vfs_unlink = unlink,
		.vfs_rename = rename,
		.vfs_fallocate = posix_fallocate,
		.vfs_ftruncate = ftruncate,
		.vfs_opendir = opendir,
		.vfs_readdir = readdir,
		.vfs_closedir = closedir,
		.vfs_mkdir = mkdir,
		.vfs_rmdir = rmdir,
		.vfs_mgr_func = polar_get_local_vfs_mgr
	},
	{
		.vfs_mount = NULL,
		.vfs_umount = NULL,
		.vfs_remount = NULL,
		.vfs_open = NULL,
		.vfs_creat = NULL,
		.vfs_close = NULL,
		.vfs_read = NULL,
		.vfs_write = NULL,
		.vfs_pread = NULL,
		.vfs_pwrite = NULL,
		.vfs_stat = NULL,
		.vfs_fstat = NULL,
		.vfs_lstat = NULL,
		.vfs_lseek = NULL,
		.vfs_access = NULL,
		.vfs_fsync = NULL,
		.vfs_unlink = NULL,
		.vfs_rename = NULL,
		.vfs_fallocate = NULL,
		.vfs_ftruncate = NULL,
		.vfs_opendir = NULL,
		.vfs_readdir = NULL,
		.vfs_closedir = NULL,
		.vfs_mkdir = NULL,
		.vfs_rmdir = NULL,
		.vfs_mgr_func = NULL
	}
};

inline int
polar_mount(void)
{
	if (polar_vfs[polar_vfs_switch].vfs_mount)
		return polar_vfs[polar_vfs_switch].vfs_mount();

	return 0;
}

inline void
polar_umount(void)
{
	if (polar_vfs[polar_vfs_switch].vfs_umount)
		return polar_vfs[polar_vfs_switch].vfs_umount();

	return;
}

inline int
polar_remount(void)
{
	int ret = 0;
	if (polar_vfs[polar_vfs_switch].vfs_remount)
		ret = polar_vfs[polar_vfs_switch].vfs_remount();
#if 0
	if (polar_enable_io_fencing && ret == 0)
	{
		/* POLAR: FATAL when shared storage is unavailable, or force to write RWID. */
		if (polar_shared_storage_is_available())
		{
			polar_hold_shared_storage(true);
			POLAR_IO_FENCING_SET_STATE(polar_io_fencing_get_instance(), POLAR_IO_FENCING_WAIT);
		}
		else
			elog(FATAL, "polardb shared storage %s is unavailable.", polar_datadir);
	}
#endif
	return ret;
}

inline int
polar_open(const char *path, int flags, mode_t mode)
{
	return polar_vfs[polar_vfs_switch].vfs_open(path, flags, mode);
}

inline int
polar_creat(const char *path, mode_t mode)
{
	return polar_vfs[polar_vfs_switch].vfs_creat(path, mode);
}

inline int
polar_close(int fd)
{
	return polar_vfs[polar_vfs_switch].vfs_close(fd);
}

inline ssize_t
polar_read(int fd, void *buf, size_t len)
{
	return polar_vfs[polar_vfs_switch].vfs_read(fd, buf, len);
}

inline ssize_t
polar_write(int fd, const void *buf, size_t len)
{
	return polar_vfs[polar_vfs_switch].vfs_write(fd, buf, len);
}

inline ssize_t
polar_pread(int fd, void *buf, size_t len, off_t offset)
{
	return polar_vfs[polar_vfs_switch].vfs_pread(fd, buf, len, offset);
}

inline ssize_t
polar_pwrite(int fd, const void *buf, size_t len, off_t offset)
{
	return polar_vfs[polar_vfs_switch].vfs_pwrite(fd, buf, len, offset);
}

/* POLAR: Replacing stat function, check all unlink when every merge code */
inline int
polar_stat(const char *path, struct stat *buf)
{
	return polar_vfs[polar_vfs_switch].vfs_stat(path, buf);
}

inline int
polar_fstat(int fd, struct stat *buf)
{
	return polar_vfs[polar_vfs_switch].vfs_fstat(fd, buf);
}

inline int
polar_lstat(const char *path, struct stat *buf)
{
	return polar_vfs[polar_vfs_switch].vfs_lstat(path, buf);
}

inline off_t
polar_lseek(int fd, off_t offset, int whence)
{
	return polar_vfs[polar_vfs_switch].vfs_lseek(fd, offset, whence);
}

inline int
polar_access(const char *path, int mode)
{
	return polar_vfs[polar_vfs_switch].vfs_access(path, mode);
}

inline int
polar_fsync(int fd)
{
	return polar_vfs[polar_vfs_switch].vfs_fsync(fd);
}

/* POLAR: Replacing unlink function, check all unlink when every merge code */
inline int
polar_unlink(const char *fname)
{
	return polar_vfs[polar_vfs_switch].vfs_unlink(fname);
}

inline int
polar_rename(const char *oldfile, const char *newfile)
{
	return polar_vfs[polar_vfs_switch].vfs_rename(oldfile, newfile);
}

inline int
polar_fallocate(int fd, off_t offset, off_t len)
{
	return polar_vfs[polar_vfs_switch].vfs_fallocate(fd, offset, len);
}

inline int
polar_ftruncate(int fd, off_t len)
{
	return polar_vfs[polar_vfs_switch].vfs_ftruncate(fd, len);
}

inline DIR	*
polar_opendir(const char *path)
{
	return polar_vfs[polar_vfs_switch].vfs_opendir(path);
}

inline struct dirent *
polar_readdir(DIR *dir)
{
	return polar_vfs[polar_vfs_switch].vfs_readdir(dir);
}

inline int
polar_closedir(DIR *dir)
{
	return polar_vfs[polar_vfs_switch].vfs_closedir(dir);
}

inline int
polar_mkdir(const char *path, mode_t mode)
{
	return polar_vfs[polar_vfs_switch].vfs_mkdir(path, mode);
}

/* POLAR: Replacing rmdir function, check all unlink when every merge code */
inline int
polar_rmdir(const char *path)
{
	return polar_vfs[polar_vfs_switch].vfs_rmdir(path);
}

static inline const vfs_mgr*
polar_get_local_vfs_mgr(const char *path)
{
	return &polar_vfs[POLAR_VFS_SWITCH_LOCAL];
}

const vfs_mgr*
polar_vfs_mgr(const char *path)
{
	if (polar_vfs[polar_vfs_switch].vfs_mgr_func)
		return (polar_vfs[polar_vfs_switch].vfs_mgr_func)(path);

	return NULL;
}

int
polar_make_pg_directory(const char *directoryName)
{
	return polar_mkdir(directoryName, pg_dir_create_mode);
}

void
polar_copydir(char *fromdir, char *todir, bool recurse)
{
	DIR		   *xldir;
	struct dirent *xlde;
	char		fromfile[MAXPGPATH * 2];
	char		tofile[MAXPGPATH * 2];

	if (polar_mkdir(todir, S_IRWXU) != 0)
		ereport(ERROR,
				(errcode_for_file_access(),
				 errmsg("could not create directory \"%s\": %m", todir)));

	xldir = polar_allocate_dir(fromdir);
	if (xldir == NULL)
		ereport(ERROR,
				(errcode_for_file_access(),
				 errmsg("could not open directory \"%s\": %m", fromdir)));

	while ((xlde = ReadDir(xldir, fromdir)) != NULL)
	{
		struct stat fst;

		CHECK_FOR_INTERRUPTS();

		if (strcmp(xlde->d_name, ".") == 0 ||
			strcmp(xlde->d_name, "..") == 0)
			continue;

		snprintf(fromfile, sizeof(fromfile), "%s/%s", fromdir, xlde->d_name);
		snprintf(tofile, sizeof(tofile), "%s/%s", todir, xlde->d_name);

		if (polar_stat(fromfile, &fst) < 0)
			ereport(ERROR,
					(errcode_for_file_access(),
					 errmsg("could not stat file \"%s\": %m", fromfile)));

		if (S_ISDIR(fst.st_mode))
		{
			if (recurse)
			{
				polar_copydir(fromfile, tofile, true);
			}
		}
		else if (S_ISREG(fst.st_mode))
			polar_copy_file(fromfile, tofile);
	}
	FreeDir(xldir);

	return;
}

void
polar_copy_file(char *fromfile, char *tofile)
{
	char	   *buffer;
	int			srcfd;
	int			dstfd;
	int			nbytes;
	off_t		offset;

#define COPY_BUF_SIZE (8 * BLCKSZ)

	buffer = palloc(COPY_BUF_SIZE);

	srcfd = polar_open_transient_file(fromfile, O_RDONLY | PG_BINARY);
	if (srcfd < 0)
		ereport(ERROR,
				(errcode_for_file_access(),
				 errmsg("could not open file \"%s\": %m", fromfile)));

	dstfd = polar_open_transient_file(tofile, O_RDWR | O_CREAT | O_EXCL | PG_BINARY);
	if (dstfd < 0)
		ereport(ERROR,
				(errcode_for_file_access(),
				 errmsg("could not create file \"%s\": %m", tofile)));

	for (offset = 0;; offset += nbytes)
	{
		CHECK_FOR_INTERRUPTS();

		pgstat_report_wait_start(WAIT_EVENT_COPY_FILE_READ);
		nbytes = polar_read(srcfd, buffer, COPY_BUF_SIZE);
		pgstat_report_wait_end();
		if (nbytes < 0)
			ereport(ERROR,
					(errcode_for_file_access(),
					 errmsg("could not read file \"%s\": %m", fromfile)));
		if (nbytes == 0)
			break;
		errno = 0;
		pgstat_report_wait_start(WAIT_EVENT_COPY_FILE_WRITE);
		if ((int) polar_write(dstfd, buffer, nbytes) != nbytes)
		{
			pgstat_report_wait_end();
			if (errno == 0)
				errno = ENOSPC;
			ereport(ERROR,
					(errcode_for_file_access(),
					 errmsg("could not write to file \"%s\": %m", tofile)));
		}
		pgstat_report_wait_end();

		polar_fsync(dstfd);
	}

	if (CloseTransientFile(dstfd))
		ereport(ERROR,
				(errcode_for_file_access(),
				 errmsg("could not close file \"%s\": %m", tofile)));

	CloseTransientFile(srcfd);

	pfree(buffer);
}

void
polar_make_file_path_level3(char *path, char *base, char *file_path)
{
	if (POLAR_FILE_IN_SHARED_STORAGE())
		snprintf(path, MAXPGPATH, "%s/%s/%s", polar_datadir, base, file_path);
	else
		snprintf(path, MAXPGPATH, "%s/%s", base, file_path);

	return;
}

void
polar_make_file_path_level2(char *path, char *file_path)
{
	if (POLAR_FILE_IN_SHARED_STORAGE())
		snprintf(path, MAXPGPATH, "%s/%s", polar_datadir, file_path);
	else
		snprintf(path, MAXPGPATH, "%s", file_path);

	return;
}

/*
 * POLAR: Extend from XLogFilePath
 */
void
polar_xLog_file_path(char *path, TimeLineID tli, XLogSegNo logSegNo, int wal_segsz_bytes)
{
	if (POLAR_FILE_IN_SHARED_STORAGE())
	{
		snprintf(path, MAXPGPATH, "%s/" XLOGDIR "/%08X%08X%08X", polar_datadir, tli,	
			 (uint32) ((logSegNo) / XLogSegmentsPerXLogId(wal_segsz_bytes)), 
			 (uint32) ((logSegNo) % XLogSegmentsPerXLogId(wal_segsz_bytes)));
	}
	else
	{
		snprintf(path, MAXPGPATH, XLOGDIR "/%08X%08X%08X", tli,	
			 (uint32) ((logSegNo) / XLogSegmentsPerXLogId(wal_segsz_bytes)), 
			 (uint32) ((logSegNo) % XLogSegmentsPerXLogId(wal_segsz_bytes)));
	}

	return;
}

/*
 * POLAR: Extend from TLHistoryFilePath
 */
void
polar_tl_history_file_path(char *path, TimeLineID tli)
{
	if (POLAR_FILE_IN_SHARED_STORAGE())
		snprintf(path, MAXPGPATH, "%s/" XLOGDIR "/%08X.history", polar_datadir, tli);
	else
		snprintf(path, MAXPGPATH, XLOGDIR "/%08X.history", tli);

	return;
}

/*
 * POLAR: StatusFilePath
 */
void
polar_status_file_path(char *path, const char *xlog, char *suffix)
{
	if (POLAR_FILE_IN_SHARED_STORAGE())
		snprintf(path, MAXPGPATH, "%s/" XLOGDIR "/archive_status/%s%s", polar_datadir, xlog, suffix);
	else
		snprintf(path, MAXPGPATH, XLOGDIR "/archive_status/%s%s", xlog, suffix);

	return;
}

/*
 * POLAR: Extend from BackupHistoryFilePath
 */
void
polar_backup_history_file_path(char *path, TimeLineID tli, XLogSegNo logSegNo, XLogRecPtr startpoint, int wal_segsz_bytes)
{
	if (POLAR_FILE_IN_SHARED_STORAGE())
	{
		snprintf(path, MAXPGPATH, "%s/" XLOGDIR "/%08X%08X%08X.%08X.backup", polar_datadir, tli,
			 (uint32) ((logSegNo) / XLogSegmentsPerXLogId(wal_segsz_bytes)),
			 (uint32) ((logSegNo) % XLogSegmentsPerXLogId(wal_segsz_bytes)),
			 (uint32) (XLogSegmentOffset((startpoint), wal_segsz_bytes)));
	}
	else
	{
		snprintf(path, MAXPGPATH, XLOGDIR "/%08X%08X%08X.%08X.backup", tli,
			 (uint32) ((logSegNo) / XLogSegmentsPerXLogId(wal_segsz_bytes)),
			 (uint32) ((logSegNo) % XLogSegmentsPerXLogId(wal_segsz_bytes)),
			 (uint32) (XLogSegmentOffset((startpoint), wal_segsz_bytes)));
	}

	return;
}

void
polar_reset_vfs_switch(void)
{
	polar_vfs_switch = POLAR_VFS_SWITCH_LOCAL;
}

