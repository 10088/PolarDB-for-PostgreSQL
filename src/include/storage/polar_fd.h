/*-------------------------------------------------------------------------
 *
 * polar_fd.h
 *
 *	  PolarDB Virtual file descriptor definitions.
 *
 * Copyright (c) 2020, Alibaba Group Holding Limited
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * src/include/storage/polar_fd.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef POLARFD_H
#define POLARFD_H

#include <dirent.h>
#include <sys/stat.h>
#include "utils/resowner.h"


/*
 * polar file system interface
 */
#define POLAR_FILE_IN_SHARED_STORAGE()				(polar_enable_shared_storage_mode)

#define POLAR_ENABLE_PWRITE()			(polar_enable_pwrite)
#define POLAR_ENABLE_PREAD()			(polar_enable_pread)

#define	AmPolarBackgroundWriterProcess()		(polar_enable_shared_storage_mode && AmBackgroundWriterProcess())

#define	POLAR_VFS_SWITCH_LOCAL		0
#define	POLAR_VFS_SWITCH_PLUGIN		1
#define POLAR_VFS_PROTOCOL_TAG	"://"

typedef enum PolarNodeType
{
	POLAR_UNKNOWN = 0,
	POLAR_MASTER = 1,
	POLAR_REPLICA,
	POLAR_STANDBY
} PolarNodeType;

extern PolarNodeType polar_local_node_type;
extern bool polar_mount_pfs_readonly_mode;
extern int	polar_vfs_switch;
extern bool polar_openfile_with_readonly_in_replica;

typedef int (*vfs_open_type) (const char *path, int flags, mode_t mode);

typedef struct vfs_mgr
{
	int			(*vfs_mount) (void);
	void		(*vfs_umount) (void);
	int			(*vfs_remount) (void);
	int			(*vfs_open) (const char *path, int flags, mode_t mode);
	int			(*vfs_creat) (const char *path, mode_t mode);
	int			(*vfs_close) (int fd);
	ssize_t		(*vfs_read) (int fd, void *buf, size_t len);
	ssize_t		(*vfs_write) (int fd, const void *buf, size_t len);
	ssize_t		(*vfs_pread) (int fd, void *buf, size_t len, off_t offset);
	ssize_t		(*vfs_pwrite) (int fd, const void *buf, size_t len, off_t offset);
	int			(*vfs_stat) (const char *path, struct stat *buf);
	int			(*vfs_fstat) (int fd, struct stat *buf);
	int			(*vfs_lstat) (const char *path, struct stat *buf);
	off_t		(*vfs_lseek) (int fd, off_t offset, int whence);
	int			(*vfs_access) (const char *path, int mode);
	int			(*vfs_fsync) (int fd);
	int			(*vfs_unlink) (const char *path);
	int			(*vfs_rename) (const char *oldpath, const char *newpath);
	int			(*vfs_fallocate) (int fd, off_t offset, off_t len);
	int			(*vfs_ftruncate) (int fd, off_t len);
	DIR		   *(*vfs_opendir) (const char *path);
	struct dirent *(*vfs_readdir) (DIR *dir);
	int			(*vfs_closedir) (DIR *dir);
	int			(*vfs_mkdir) (const char *path, mode_t mode);
	int			(*vfs_rmdir) (const char *path);
	const struct vfs_mgr *(*vfs_mgr_func) (const char *path);
} vfs_mgr;

extern vfs_mgr polar_vfs[];

extern int	polar_mount(void);
extern void polar_umount(void);
extern int	polar_remount(void);

extern const vfs_mgr *polar_vfs_mgr(const char *path);

extern int	polar_open(const char *path, int flags, mode_t mode);
extern int	polar_creat(const char *path, mode_t mode);
extern int	polar_close(int fd);

extern ssize_t polar_read(int fd, void *buf, size_t len);
extern ssize_t polar_write(int fd, const void *buf, size_t len);
extern ssize_t polar_pread(int fd, void *buf, size_t len, off_t offset);
extern ssize_t polar_pwrite(int fd, const void *buf, size_t len, off_t offset);

extern int	polar_stat(const char *path, struct stat *buf);
extern int	polar_fstat(int fd, struct stat *buf);
extern int	polar_lstat(const char *path, struct stat *buf);
extern off_t polar_lseek(int fd, off_t offset, int whence);
extern int	polar_fallocate(int fd, off_t offset, off_t len);
extern int	polar_ftruncate(int fd, off_t len);
extern int	polar_access(const char *path, int mode);
extern int	polar_fsync(int fd);
extern int	polar_unlink(const char *fname);
extern int	polar_rename(const char *oldfile, const char *newfile);

extern DIR *polar_opendir(const char *path);
extern struct dirent *polar_readdir(DIR *dir);
extern int	polar_closedir(DIR *dir);
extern int	polar_mkdir(const char *path, mode_t mode);
extern int	polar_rmdir(const char *path);
extern int	polar_make_pg_directory(const char *directoryName);

extern void polar_copy_file(char *fromfile, char *tofile, bool skiperr);
extern void polar_copydir(char *fromdir, char *todir, bool recurse, bool clean, bool skiperr);

extern void polar_make_file_path_level3(char *path, char *base, char *file_path);
extern void polar_make_file_path_level2(char *path, char *file_path);

extern void polar_reset_vfs_switch(void);
extern void polar_init_node_type(void);
extern PolarNodeType polar_node_type_by_file(void);
extern struct dirent *polar_read_dir_ext(DIR *dir, const char *dirname, int elevel, int *err);
const char *polar_path_remove_protocol(const char *path);
#endif
