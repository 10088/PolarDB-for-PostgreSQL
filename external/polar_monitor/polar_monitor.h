/*-------------------------------------------------------------------------
 *
 * procstat.h
 *
 *	Copyright (c) 2020, Alibaba.inc
 *
 * IDENTIFICATION
 *		external/polar_monitor/polar_monitor.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef POLAR_MONIROT_H
#define POLAR_MONIROT_H

#include "postgres.h"

#include "access/htup_details.h"
#include "access/slru.h"
#include "access/xact.h"
#include "access/xlog.h"
#include "catalog/pg_authid.h"
#include "catalog/pg_type.h"
#include "funcapi.h"
#include "pgstat.h"
#include "libpq/libpq-be.h"
#include "miscadmin.h"
#include "postmaster/bgworker.h"
#include "procstat.h"
#include "replication/slot.h"
#include "replication/walreceiver.h"
#include "replication/walsender.h"
#include "storage/buf_internals.h"
#include "storage/bufmgr.h"
#include "storage/pg_shmem.h"
#include "storage/polar_io_stat.h"
#include "storage/procarray.h"
#include "utils/builtins.h"
#include "utils/guc.h"
#include "utils/pg_lsn.h"
#include "utils/timestamp.h"

extern Datum polar_stat_process(PG_FUNCTION_ARGS);
extern Datum polar_stat_io_info(PG_FUNCTION_ARGS);
extern Datum polar_io_latency_info(PG_FUNCTION_ARGS);

#endif	
