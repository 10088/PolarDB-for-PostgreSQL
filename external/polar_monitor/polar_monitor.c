/*-------------------------------------------------------------------------
 *
 * polar_monitor.c
 *  display some information of polardb
 *   
 * Copyright (c) 2018, Alibaba Group Holding Limited
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.*
 * 
 * IDENTIFICATION
 *  external/polar_monitor/polar_monitor.c
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/xlog.h"
#include "funcapi.h"
#include "storage/polar_fd.h"
#include "utils/builtins.h"
#include "utils/pg_lsn.h"

PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(polar_consistent_lsn);

Datum
polar_consistent_lsn(PG_FUNCTION_ARGS)
{
	XLogRecPtr cosistent_lsn;

	cosistent_lsn = polar_get_consistent_lsn();
	PG_RETURN_LSN(cosistent_lsn);
}

PG_FUNCTION_INFO_V1(polar_oldest_apply_lsn);

Datum
polar_oldest_apply_lsn(PG_FUNCTION_ARGS)
{
	if (RecoveryInProgress())
		ereport(ERROR,
				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
				 errmsg("recovery is in progress"),
				    errhint("WAL control functions cannot be executed during recovery.")));

	PG_RETURN_LSN(polar_get_oldest_applied_lsn());
}

PG_FUNCTION_INFO_V1(polar_get_node_type);
Datum
polar_get_node_type(PG_FUNCTION_ARGS)
{
	char	*mode;
	PolarNodeType node_type = polar_node_type();

	switch (node_type)
	{
		case POLAR_MASTER:
			mode = "master";
			break;
		case POLAR_REPLICA:
			mode = "replica";
			break;
		case POLAR_STANDBY:
			mode = "standby";
			break;
		default:
			mode = "unknown";
	}

	PG_RETURN_TEXT_P(cstring_to_text(mode));
}