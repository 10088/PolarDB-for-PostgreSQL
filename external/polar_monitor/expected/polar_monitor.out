-- Test access privileges
--
--setup
create extension polar_monitor;
-- buffer monitor
select COUNT(*) >= 0 AS result from polar_normal_buffercache;
 result 
--------
 t
(1 row)

select COUNT(*) >= 0 AS result from polar_copy_buffercache;
 result 
--------
 t
(1 row)

select COUNT(polar_flushlist()) >= 0 As result;
 result 
--------
 t
(1 row)

select COUNT(polar_cbuf()) >= 0 As result;
 result 
--------
 t
(1 row)

--cleanup
drop extension polar_monitor;
